### PR TITLE
sidecar: add support for http_proxy protocol

### PR DESCRIPTION
**Pull request by ramaraochavali**
Adds support for http_proxy protocol as generic egress proxy.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@rshriram `http_proxy` protocol support. Once this is merged, I will update api documentation.

File: tests/e2e/tests/pilot/testdata/networking/v1alpha3/sidecar-scope-ns1-ns2.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ spec:
   egress:
   - port:
       number: 23145
-      protocol: TCP
+      protocol: http_proxy
```
Comment (howardjohn) [inline tests/e2e/tests/pilot/testdata/networking/v1alpha3/sidecar-scope-ns1-ns2.yaml]: this seems very confusing given that http-proxy means something completely different in the "port name magic". Is this field not an enum? Why are we adding a new protocol?
also this looks like it has to be lower case

**Comment by ramaraochavali** [comment]
It is still in WIP. I will add it to the enum and make
It upper case etc. But the idea is to specify it as http proxy port so that
it acts as a outbound proxy.

File: pilot/pkg/model/sidecar.go [hunk 2]
```diff
@@ -517,9 +518,11 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
...
 
 func matchingServices(importedHosts []host.Name, service *Service, ilw *IstioEgressListenerWrapper) []*Service {
-	// If a listener is defined with port, we should match services with port.
-	// If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
-	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil && ilw.IstioListener.Port.GetNumber() != 0
+	// If a listener is defined with port, we should match services with port except in the following cases.
+	//  - If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
+	//  - If Port's protocol is proxy protocol in which case the egress listener is used as generic egress http proxy.
+	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil &&
+		ilw.IstioListener.Port.GetNumber() != 0 && ilw.IstioListener.Port.Protocol != proxyProtocol
```
Comment (rshriram) [inline pilot/pkg/model/sidecar.go]: Why not add it to protocol list ?

Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go]: I am thinking of adding - just pushed to see what CI thinks

File: pilot/pkg/model/sidecar.go [hunk 3]
```diff
@@ -517,9 +518,11 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
 }
 
 func matchingServices(importedHosts []host.Name, service *Service, ilw *IstioEgressListenerWrapper) []*Service {
-	// If a listener is defined with port, we should match services with port.
-	// If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
-	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil && ilw.IstioListener.Port.GetNumber() != 0
+	// If a listener is defined with port, we should match services with port except in the following cases.
+	//  - If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go]: > If an unix domain socket is given as a port

How can i comprehend this?

File: pilot/pkg/model/sidecar.go [hunk 4]
```diff
@@ -517,9 +518,11 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
...
+	//  - If Port's protocol is proxy protocol in which case the egress listener is used as generic egress http proxy.
+	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil &&
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go]: nit: you can remove ` ilw.IstioListener.Port != nil` check, it's safe

File: pilot/pkg/model/sidecar.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go]: I think that comment is redundant. I will remove that - basically if a port has some number we should proceed with matching - does not matter whether it is a unix domain port or not.

**Comment by ramaraochavali** [comment]
@rshriram  added to protocol enum and CI also seems happy. Can you PTAL and see if any other changes are needed?

**Comment by ramaraochavali** [comment]
@rshriram Can you PTAL when you get some time? Not urgent.

File: pilot/pkg/model/sidecar.go [hunk 5]
```diff
@@ -517,9 +518,10 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
 }
 
 func matchingServices(importedHosts []host.Name, service *Service, ilw *IstioEgressListenerWrapper) []*Service {
-	// If a listener is defined with port, we should match services with port.
-	// If an unix domain socket is given as a port, we should not match by port and include services based on hosts.
-	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port != nil && ilw.IstioListener.Port.GetNumber() != 0
+	// If a listener is defined with a port, we should match services with port except in the following case.
+	//  - If Port's protocol is proxy protocol(HTTP_PROXY) in which case the egress listener is used as generic egress http proxy.
+	needsPortMatch := ilw.IstioListener != nil && ilw.IstioListener.Port.GetNumber() != 0 &&
+		protocol.Parse(ilw.IstioListener.Port.Protocol) != protocol.HTTP_PROXY
```
Comment (rshriram) [inline pilot/pkg/model/sidecar.go]: where is HTTP_PROXY declared? I dont see it in that package

File: pkg/config/protocol/instance.go [hunk 6]
```diff
@@ -63,6 +66,8 @@ func Parse(s string) Instance {
 		return GRPCWeb
 	case "http":
 		return HTTP
+	case "http_proxy":
+		return HTTP_PROXY
```
Comment (rshriram) [inline pkg/config/protocol/instance.go]: shouldnt the is HTTP functions be updated as well ?

Comment (ramaraochavali) [reply pkg/config/protocol/instance.go]: Updated isHttp function. Not sure if we have to update other - I do not think so. But please let me know if I missed any thing.

**Comment by ramaraochavali** [comment]
@rshriram can you PTAL at this when you get time?

**Pull request merged**